sockeye/inference.py

```diff
@@ -1043,28 +1043,67 @@ def _beam_search(self,
             if self.restrict_lexicon:
                 best_word_indices[:] = vocab_slice_ids.take(best_word_indices)
 
-            # (4) get hypotheses and their properties for beam_size winning hypotheses (ascending)
+            # (4) Normalize the scores of newly finished hypotheses
+            newly_finished = (best_word_indices == self.vocab_target[C.EOS_SYMBOL]).expand_dims(axis=1)
+            scores_accumulated = mx.nd.where(newly_finished, scores_accumulated / self.length_penalty(lengths), scores_accumulated)
+
+            # (5) Prune out low-probability hypotheses
+            for sentno in range(self.batch_size):
```

Comment:
how much runtime does this new feature add? How is speed affected if you don't prune (the baseline setting)?

Reply:
On GPUs it's the same or a little faster with no pruning; on CPUs it's maybe a 10–20% speedup. There's more work in some places (e.g., here), but you also save a lot in other places. Being able to prune the hypothesis list puts in place a more general framework where we could do more advanced filtering.